R/guide-.r

```diff
@@ -0,0 +1,391 @@
+      return(out)
+    }
+
+    # Build grobs
+    grob_title  <- self$build_title(params$title, elems, params)
+    grob_labels <- self$build_labels(key, elems, params)
+    grob_ticks  <- self$build_ticks(key, elems, params)
+    grob_decor  <- self$build_decor(self$decor, grob_ticks, elems, params)
+    grobs <- list(
+      title = grob_title,
+      label = grob_labels,
+      ticks = grob_ticks,
+      decor = grob_decor
+    )
+
+    # Arrange and assemble grobs
+    sizes  <- self$measure_grobs(grobs, params, elems)
+    layout <- self$arrange_layout(key, sizes, params)
+    self$assemble_drawing(grobs, layout, sizes, params)
+  },
+
+  # Makes measurements of grobs that can be used in the layout or assembly
+  # stages of guide drawing.
+  measure_grobs = function(grobs, params) {
+    return(invisible())
+  },
+
+  # Takes care of where grobs should be added to the output gtable.
+  arrange_layout = function(key, sizes, params) {
+    return(invisible())
+  },
+
+  # Combines grobs into a single gtable.
+  assemble_drawing = function(grobs, layout, sizes, params) {
+    zeroGrob()
+  },
+
+  # Renders the guide title
+  build_title = function(label, elements, params) {
+    ggname(
+      element_grob(
+        elements$title,
+        label    = label,
+        margin_x = TRUE,
+        margin_y = TRUE
+      ),
+      "guide.title"
+    )
+  },
+
+  # Renders the guide labels
+  # TODO: See if we can generalise label drawing for many guides
+  build_labels = function(key, elements, params) {
+    zeroGrob()
+  },
+
+  # Renders 'decor', which can have different meanings for different guides.
+  # Ticks are provided because they may need to be combined with decor.
+  build_decor = function(decor, ticks, elements, params) {
+    zeroGrob()
+  },
+
+  # Renders tickmarks
+  build_ticks = function(key, elements, params, position = params$position) {
+    breaks   <- key[[params$aes]] %||% key
+    n_breaks <- length(breaks)
+
+    # Early exit if there are no breaks
+    if (n_breaks < 1) {
+      return(zeroGrob())
+    }
+
+    tick_len <- rep(elements$ticks_length %||% unit(0.2, "npc"),
+                    length.out = n_breaks)
+
+    # Resolve mark
+    mark <- unit(rep(breaks, each = 2), "npc")
+
+    # Resolve ticks
+    pos <- unname(c(top = 1, bottom = 0, left = 0, right = 1)[position])
+    dir <- -2 * pos + 1
+    pos <- unit(rep(pos, 2 * n_breaks), "npc")
+    dir <- rep(vctrs::vec_interleave(0, dir), n_breaks) * tick_len
+    tick <- pos + dir
+
+    # Build grob
+    bidi_element_grob(
+      elements$ticks,
+      x = tick, y = mark,
+      id.lengths = rep(2, n_breaks),
+      flip = position %in% c("top", "bottom")
+    )
+  },
+
+  draw_early_exit = function(self, params, elements) {
+    zeroGrob()
+  }
+)
+
+# Helper function that may facilitate bidirectional theme elements by
+# flipping x/y related arguments to `element_grob()`
+bidi_element_grob = function(..., flip = FALSE) {
```

Comment:
small point but _bidi_ mostly refers to left-to-right and right-to-left text, rather than a x-to-y translation